### PR TITLE
feat(gnoweb): rework & Implement new gnoweb design

### PR DESCRIPTION
**Pull request by gfanton**
address #3191

Reworking the `gnoweb` package:  
-  Implement `gnoweb` new interface design(cc @alexiscolin).  
-  Move Markdown rendering to the server to enhance speed and security. This change also simplifies the implementation of new components, making it more standardized as a Go library.  
-  Aim to keep dependencies minimal, using only `goldmark` for Markdown and `chroma` for code highlighting, with almost no (in)direct dependencies.  
-  Transition to Tailwind for simplicity and maintainability.  
-  Retain all features from the previous `gnoweb` iteration.  

### Preview
- Home
Screenshot 2024-11-25 at 19 39 54

- Source
Screenshot 2024-11-25 at 19 41 25

- Docs
Screenshot 2024-11-25 at 19 45 16

### TODO: 
- [x] port and adapt all previous tests to ensure compatibility (it should not take too long)
- [x] Some cleanup and restructuring + linting.

### CODE REVIEW

**Comment by Gno2D2** [comment]
I'm a bot that assists the Gno Core team in maintaining this repository. My role is to ensure that contributors understand and follow our guidelines, helping to streamline the development process.

The following requirements must be fulfilled before a pull request can be merged.
Some requirement checks are automated and can be verified by the CI, while others need manual verification by a staff member.

These requirements are defined in this [configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

## Automated Checks

 🟢 Maintainers must be able to edit this pull request
 🔴 The pull request head branch must be up-to-date with its base


## Manual Checks

*No manual checks match this pull request.*

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>The pull request head branch must be up-to-date with its base</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Head branch (gfanton:feat/gnoweb-rework) is up to date with base (master): behind by 24 / ahead by 12

```
</blockquote></details>

</blockquote></details>



</blockquote></details>

File: gno.land/pkg/gnoweb/app.go
Comment (leohhhn) [inline gno.land/pkg/gnoweb/app.go:47]:
```suggestion
		return nil, fmt.Errorf("unable to create http client: %w", err)
```

File: gno.land/pkg/gnoweb/alias.go
Comment (leohhhn) [inline gno.land/pkg/gnoweb/alias.go:28]: We can remove this I think

Comment (alexiscolin) [reply gno.land/pkg/gnoweb/alias.go:28]: At some point, we should add a param to specify whether the redirects are permanent or temporary.

**Comment by leohhhn** [comment]
Hey, I took a look at the draft PR version of gnoweb, and here are a few things I found that are bugs/could be improved.

- Relative vs absolute links: ~for some reason, the current gnoweb does not differentiate between relative and absolute links properly (see current homepage).~ Works now, didn't work yesterday. I'll test this out again later.
- If `$help&func=XYZ`, only the `XYZ` function should appear, or should be prominent in some other way to the user
- TOC + $help page are weird, ie `/r/morgan/guestbook$help&func=Sign#func-Sign` 
	- in this case (the $help page), i think we should disable TOC links and just use the gnoweb query arg to "select" the function in question
- Adding a "Back" button would be amazing!!
- p/ pages should render the package name, and the list of files. gnoweb++ right now tries to query "Render()" which shouldn't be the case
- possible improvements could be:
	- show doc.gno if exists, show it
	- a godoc style index/docs showing for packages
- Clicking on the “/r” in the search bar leads you to a “not found”. Right now I think we just handle it somehow, ie redirect to home page
- When a realm render doesn't have much content, you need to scroll down to see the footer. Maybe even make the footer a bit taller :slightly_smiling_face:
- `Render()` in realms is optional. This means that we should handle this case specifically on the frontend, ie "this realm does not have a render function", but also on the back. Currently, an error is displayed, while this technically isn't an error.
Screenshot 2024-12-06 at 10 18 59

Also, let's use lowercase for our tools in general. ie `gnoweb` instead of `GnoWeb`.

File: gno.land/pkg/gnoweb/components/index.gohtml
Comment (leohhhn) [inline gno.land/pkg/gnoweb/components/index.gohtml:132]:
```suggestion
                <li><a href="https://github.com/gnolang/gno">GitHub</a></li>
```

Comment (gfanton) [reply gno.land/pkg/gnoweb/components/index.gohtml]: c20a99b408a8dbefcd11d2a01b0b474e8158ff80

File: gno.land/pkg/gnoweb/alias.go
Comment (gfanton) [reply gno.land/pkg/gnoweb/alias.go]: removed

File: gno.land/cmd/gnoweb/main.go
Comment (leohhhn) [inline gno.land/cmd/gnoweb/main.go:23]: Just curious; why `bind`?

File: gno.land/pkg/gnoweb/url.go
Comment (leohhhn) [inline gno.land/pkg/gnoweb/url.go:23]: I think `domain` would be the better name here, matching what we have currently in gno.land (ie `ChainDomain`).

File: gno.land/pkg/gnoweb/url_test.go
Comment (leohhhn) [inline gno.land/pkg/gnoweb/url_test.go:91]: So, the gnoweb query and the realm query need to be in a particular order, ie 
`pkgpath:renderpath$tz=Europe/Paris?realmquerykey=value&newvalue=1`?

gnoweb query needs to be before the realm query?

File: gno.land/pkg/gnoweb/url_test.go
Comment (leohhhn) [inline gno.land/pkg/gnoweb/url_test.go:106]: Nit: I guess maybe a more intuitive naming would be 
`pkgpath:renderpath?renderquery=123&anotherrenderquery=something`

Not sure though.

File: gno.land/pkg/gnoweb/url_test.go
Comment (leohhhn) [inline gno.land/pkg/gnoweb/url_test.go:109]: Maybe find a quick way to autogenerate/fuzz these?

File: gno.land/pkg/gnoweb/webclient.go
Comment (leohhhn) [inline gno.land/pkg/gnoweb/webclient.go:38]: Nit: Arguably the whole thing is the pkgpath, including the chain domain.

File: gno.land/pkg/gnoweb/webclient.go
Comment (leohhhn) [inline gno.land/pkg/gnoweb/webclient.go:61]: Might take this up

https://github.com/gnolang/gno/issues/3305

**Review by leohhhn — APPROVED** [review]
Took a general look, left some minor comments. Front looks good.

File: gno.land/cmd/gnoweb/main.go
Comment (gfanton) [reply gno.land/cmd/gnoweb/main.go:23]: It's common practice to use `bind` as a flag name because it literally binds a listening socket to a specific address and port

File: gno.land/pkg/gnoweb/url_test.go
Comment (gfanton) [reply gno.land/pkg/gnoweb/url_test.go:91]: Yes, the order matters. More specifically, we don't want to have webquery within the query because it would make the URL non-compliant with standard URL format, and most common REST tools wouldn't be able to correctly manipulate this kind of URL, i.e., Postman. It would also make parsing and error management more complex to implement.
c.f https://github.com/gnolang/gno/issues/3162

File: gno.land/pkg/gnoweb/url_test.go
Comment (gfanton) [reply gno.land/pkg/gnoweb/url_test.go:106]: do you mean just for this specific test right?

**Review by thehowl — COMMENTED** [review]
okay forgot to comment these, doing a proper review next up

File: gno.land/pkg/markdown/toc.go
Comment (thehowl) [reply gno.land/pkg/markdown/toc.go:1]: after discussion, think this should be gno.land/pkg/gnoweb/markdown

we can move to be more generic in the future and then to gno.land/pkg/markdown, but i suggest you focus on the specific case for now

**Review by thehowl — APPROVED** [review]
pr on fork incoming

File: gno.land/pkg/gnoweb/handler.go
Comment (thehowl) [inline gno.land/pkg/gnoweb/handler.go:1]: General comment: we never set a page title anywhere. Can we configure page titles? :)

**Comment by thehowl** [comment]
other general comment, but I'm sure you thought about it:

it would be cool if the "search bar" (where the realm path is) acted kind of as a "command palette", and as such allowed to move quickly to help and source, and to move to other realms.

I imagine this is somewhat envisioned/planned. I'm looking forward :)

File: gno.land/pkg/gnoweb/handler.go
Comment (gfanton) [reply gno.land/pkg/gnoweb/handler.go]: db88955594c2965515fbb84210f2059ebe336d20

File: contribs/gnodev/cmd/gnodev/setup_web.go
Comment (moul) [inline contribs/gnodev/cmd/gnodev/setup_web.go:9]:
```suggestion
 "github.com/gnolang/gno/gno.land/pkg/gnoweb"
```

File: gno.land/cmd/gnoweb/README.md
Comment (moul) [inline gno.land/cmd/gnoweb/README.md:1]: Keep a README in this folder, but feel free to mention that people can view previews on gno.land or run it with gnodev, linking to the gnodev folder.

File: gno.land/cmd/gnoweb/main_test.go
Comment (moul) [inline gno.land/cmd/gnoweb/main_test.go:1]: Please keep it this way: 1. to facilitate the addition of second tests later, and 2. to ensure that `go test` does not skip this folder and passes the basic syntax checks.

File: gno.land/pkg/gnoweb/components/breadcrumb.gohtml
Comment (moul) [inline gno.land/pkg/gnoweb/components/breadcrumb.gohtml:1]: Can we consider not committing those generated files and ensure that each install command runs `go generate`?

File: gno.land/pkg/gnoweb/Makefile
Comment (thehowl) [inline gno.land/pkg/gnoweb/Makefile:45]: should have a CI workflow to ensure it's always up-to-date

**Comment by leohhhn** [comment]
@alexiscolin another small bug; with orderedlists the number does not show up:

Screenshot 2024-12-10 at 18 52 36

File: misc/deployments/staging.gno.land/docker-compose.yml
Comment (sw360cab) [inline misc/deployments/staging.gno.land/docker-compose.yml:1]: Since faucet config is removed from config, would it not be accessible from Gnoweb anymore?
Which is the entrypoint of Faucet so far? Because `CAPTCHA_SITE_KEY` is not set anywhere else, and is the one intended to be used by a _Gnofaucet frontend_.

cc: @zivkovicmilos

File: gno.land/cmd/gnoweb/main.go
Comment (gfanton) [reply gno.land/cmd/gnoweb/main.go:26]: Keeping the original behaviors of `gnoweb`, it has also been added recently to gnodev by @moul, so I believe we still need to keep this around as an option.

File: gno.land/pkg/gnoweb/Makefile
Comment (gfanton) [reply gno.land/pkg/gnoweb/Makefile:1]: yes definitely

**Comment by leohhhn** [comment]
Hey, please confirm with me directly before merging and updating the testX/portal loop `gnoweb` images; we will be sharing an announcement on social media about the new design revamp and we need to coordinate it. cc @sw360cab

File: gno.land/cmd/gnoweb/main_test.go
Comment (gfanton) [reply gno.land/cmd/gnoweb/main_test.go:1]: I cannot _exactly_ keep it this way as I'm using the `command` package. But I found another way to implement it (I personally think it's even better as I test the whole setup of the main).

File: gno.land/pkg/gnoweb/components/breadcrumb.gohtml
Comment (gfanton) [reply gno.land/pkg/gnoweb/components/breadcrumb.gohtml:1]: These files are not generated; they are simply Go HTML templates. That aside, we need to push generated files, as they are mostly CSS and JS, and we don't want users to depend on Node to install gnodev or gnoweb. I've added a CI rule to check if generated files are correctly synced.

File: gno.land/pkg/gnoweb/README.md
Comment (gfanton) [reply gno.land/pkg/gnoweb/README.md:69]: cleaned

File: gno.land/pkg/gnoweb/README.md
Comment (gfanton) [reply gno.land/pkg/gnoweb/README.md:3]: I've updated it with a minimal description

File: gno.land/pkg/gnoweb/README.md
Comment (gfanton) [reply gno.land/pkg/gnoweb/README.md:1]: done

File: gno.land/pkg/markdown/toc.go
Comment (gfanton) [reply gno.land/pkg/markdown/toc.go:1]: done

File: gno.land/pkg/gnoweb/app.go
Comment (gfanton) [reply gno.land/pkg/gnoweb/app.go]: done

File: misc/deployments/staging.gno.land/docker-compose.yml
Comment (gfanton) [reply misc/deployments/staging.gno.land/docker-compose.yml:1]: After some internal discussion `/faucet` will now simply redirect to `faucet-hub`

**Review by thehowl — APPROVED** [review]

**Pull request merged**
